benchmark/CMakeLists.txt

```diff
@@ -1,283 +1,28 @@
-)
-
-set(
-    CholQRCP_inner_speed_cxx
-    CholQRCP_inner_speed.cc
-)
-add_executable(
-    CholQRCP_inner_speed ${CholQRCP_inner_speed_cxx}
-)
-target_compile_options(
-    CholQRCP_inner_speed PRIVATE "-g"
-)
-set(
-    RandLAPACK_include_dirs
-    ${RandLAPACK_DIR}/../../include
-    ${RandLAPACK_DIR}/../../include/RandLAPACK
-)
-target_include_directories(
-    CholQRCP_inner_speed PUBLIC ${Random123_DIR} ${RandLAPACK_include_dirs}
-)
-set(
-    RandLAPACK_lib_file
-    ${RandLAPACK_DIR}/../../lib/libRandLAPACK.a
-)
-target_link_libraries(
-    CholQRCP_inner_speed PUBLIC ${RandLAPACK_lib_file} RandBLAS blaspp lapackpp
-)
-
-set(
-    CholQRCP_inner_comp_cxx
-    CholQRCP_inner_comp.cc
-)
-add_executable(
-    CholQRCP_inner_comp ${CholQRCP_inner_comp_cxx}
-)
-target_compile_options(
-    CholQRCP_inner_comp PRIVATE "-g"
-)
-set(
-    RandLAPACK_include_dirs
-    ${RandLAPACK_DIR}/../../include
-    ${RandLAPACK_DIR}/../../include/RandLAPACK
-)
-target_include_directories(
-    CholQRCP_inner_comp PUBLIC ${Random123_DIR} ${RandLAPACK_include_dirs}
-)
-set(
-    RandLAPACK_lib_file
-    ${RandLAPACK_DIR}/../../lib/libRandLAPACK.a
-)
-target_link_libraries(
-    CholQRCP_inner_comp PUBLIC ${RandLAPACK_lib_file} RandBLAS blaspp lapackpp
-)
-
-set(
-    CholQRCP_inner_speed_process_cxx
-    CholQRCP_inner_speed_process.cc
-)
-add_executable(
-    CholQRCP_inner_speed_process ${CholQRCP_inner_speed_process_cxx}
-)
-target_compile_options(
-    CholQRCP_inner_speed_process PRIVATE "-g"
-)
-set(
-    RandLAPACK_include_dirs
-    ${RandLAPACK_DIR}/../../include
-    ${RandLAPACK_DIR}/../../include/RandLAPACK
-)
-target_include_directories(
-    CholQRCP_inner_speed_process PUBLIC ${Random123_DIR} ${RandLAPACK_include_dirs}
-)
-set(
-    RandLAPACK_lib_file
-    ${RandLAPACK_DIR}/../../lib/libRandLAPACK.a
-)
-target_link_libraries(
-    CholQRCP_inner_speed_process PUBLIC ${RandLAPACK_lib_file} RandBLAS blaspp lapackpp
-)
-
-set(
-    CholQRCP_embedding_speed_effect_cxx
-    CholQRCP_embedding_speed_effect.cc
-)
-add_executable(
-    CholQRCP_embedding_speed_effect ${CholQRCP_embedding_speed_effect_cxx}
-)
-target_compile_options(
-    CholQRCP_embedding_speed_effect PRIVATE "-g"
-)
-set(
-    RandLAPACK_include_dirs
-    ${RandLAPACK_DIR}/../../include
-    ${RandLAPACK_DIR}/../../include/RandLAPACK
-)
-target_include_directories(
-    CholQRCP_embedding_speed_effect PUBLIC ${Random123_DIR} ${RandLAPACK_include_dirs}
-)
-set(
-    RandLAPACK_lib_file
-    ${RandLAPACK_DIR}/../../lib/libRandLAPACK.a
-)
-target_link_libraries(
-    CholQRCP_embedding_speed_effect PUBLIC ${RandLAPACK_lib_file} RandBLAS blaspp lapackpp
-)
+function(add_benchmark)
+    set(OPTS)
+    set(NVPO NAME)
+    set(MVO CXX_SOURCES LINK_LIBS)
+    cmake_parse_arguments(PARSE_ARGV 0 TGT "${OPTS}" "${NVPO}" "${MVO}")
+    add_executable(${TGT_NAME} ${TGT_CXX_SOURCES})
+    target_compile_options(${TGT_NAME} PRIVATE "-O3")
```

Comment:
This is probably a bad idea. CMake sets -O3 on release builds, so it is superfluous if one selects a release build, which is a good idea since there are other flags that need to be set as well. Also, what if we are trying top debug the code? -O3 will make it hard to single step through the sources because of inlining. In other words -O3 should not be set for debug build.

Reply:
Thanks for the insight @burlen. Max and I discussed making this change for benchmarks specifically because benchmarks should not be built in debug mode. Debugging might be needed when the benchmark is first written, but it shouldn't be needed once the benchmark has been merged into main.

Based on that, do you think using -O3 here is okay, or would you still object?

Reply:
I think of it this way: Data from benchmarks is not useful unless it was compiled with optimizations. Benchmarks should be built in debug mode when debugging them, and release mode when profiling.

This doesn't ensure that other parts of the library will end up with the correct flags. Not so much of an issue now that we're header only.

You should also at least have -DNDEBUG. Other compilers have different meanings for there flags, e.x. with intel -O2 is closer to gcc's -O3. You'd need quite a bit more here than what you've got to do this right.

If/When we need to debug these in the future, it will be confusing, and we'll have to know to modify this line of cmake code. We can save our future selves time now.

If you're worried about folks making timings on debug builds you can add something like
```C++
#if !defined(NDEBUG)
    std::cerr << "WARNING: Benchmark compiled in debug mode, timing is invalid!" << std::endl;
#endif
```
to you main function, or in the timing outputs. Then at least there is some hint.

Reply:
Thanks @burlen, that makes sense. I love the suggested warning message for the benchmarks!

@TeachRaccooon, I would revert this default change then. How about you create a README about compiling and running the benchmarks? It doesn't need to be perfect since it will end up in a different repo soon enough, but it would be a good starting point.